### ISSUE TITLE
Sometimes crashes with an IndexOutOfBoundsException if there's too many log messages.

### ISSUE THREAD

**Issue by codebymikey** [jraska/Console#9]
The console widget crashes my app whenever I have too many log messages call.
The stack-trace is as follows:

```
FATAL EXCEPTION: main
    Process: com.example.sample.debug, PID: 9708
    java.lang.IndexOutOfBoundsException: getChars (0 ... 16098) ends beyond length 16001
        at android.text.SpannableStringBuilder.checkRange(SpannableStringBuilder.java:1265)
        at android.text.SpannableStringBuilder.getChars(SpannableStringBuilder.java:1147)
        at android.text.TextUtils.getChars(TextUtils.java:89)
        at android.text.SpannableStringBuilder.<init>(SpannableStringBuilder.java:66)
        at android.text.SpannableStringBuilder.<init>(SpannableStringBuilder.java:50)
        at android.text.Editable$Factory.newEditable(Editable.java:143)
        at android.widget.TextView.setText(TextView.java:4391)
        at android.widget.TextView.setText(TextView.java:4332)
        at android.widget.TextView.setText(TextView.java:4307)
        at com.jraska.console.ConsoleBuffer.printTo(ConsoleBuffer.java:59)
        at com.jraska.console.Console.printBuffer(Console.java:273)
        at com.jraska.console.Console.printScroll(Console.java:268)
        at com.jraska.console.Console.runBufferPrint(Console.java:304)
        at com.jraska.console.Console.access$300(Console.java:29)
        at com.jraska.console.Console$BufferPrintRunnable.run(Console.java:318)
        at android.os.Handler.handleCallback(Handler.java:751)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6077)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
```

I'd recommend either:
1. Debouncing the `printBuffer()` calls to every 2 seconds (configurable by the user), or
2. Moving to a simple `RecyclerView`, that way you don't have a massive `TextView` object with lots of `Spannable`s, and only show logs that are visible. Which will be more memory efficient.

Option 2 is what I'd prefer going forward, but I'm aware it's likely a big refactoring job.

**Comment by jraska** [comment]
Hi, 

Thank you for reporting this and for ideas, I see concurrency issue there. I will try to reproduce it and fix it as soon as possible.

**Comment by jraska** [comment]
I reproduced the issue and I think I fixed that. Please check version 0.4.3 and let me know if it helped.

`compile 'com.jraska:console:0.4.3'`

**Comment by codebymikey** [comment]
Thanks again for the relatively quick solution! I've been pretty busy lately.

Yes, adding the `synchronized` block seems to work, but results in the app becoming virtually unusable (functional but incredibly slow) because it's blocking the main thread while adding the items one by one.

I think the `debounce` solution (which should probably be optional and it's interval duration be configurable) should hopefully solve this.

It'd also be nice if the `ConsoleBuffer` class was made non-final so that people who need more flexibility over the buffer and how it works should have the option to do so.

This is the block of code I was messing around with which triggers the thread blocking:

```
Observable.zip(Observable.range(1, 100000), Observable.interval(1, TimeUnit.NANOSECONDS), new Func2<Integer, Long, Integer>() {
        @Override
    public Integer call(Integer integer, Long aLong) {
        return integer;
    }
})
.subscribeOn(Schedulers.computation())
.subscribe(new Action1<Integer>() {
    @Override
    public void call(Integer integer) {
        Timber.v("%s", integer);
    }
});
```

**Comment by jraska** [comment]
Hi, thank you very much for the suggestion. I understand what you mean and solution with RxJava sounds nice. 
However, I don't want to bring RxJava dependency to Console and it will take a while to find time for implementing something like that on my own.

Making these things configurable can be an option, but I am not keen to open the interface too much, since I wanted to keep it as simple as possible and again it will require some work which I am not sure is worth to do now.

I have a suggestion what you can try now which may solve your issue without update of Console which will take a while.

You can implement debouncing or buffering on your own and concatenate buffered messages before you log them.

```
Observable.range(1, 10000)
        .buffer(2, TimeUnit.SECONDS, 100)
        .subscribe(new Action1<List<Integer>>() {
          @Override public void call(List<Integer> integers) {
            String message = concatValues(integers);
            Timber.v(message);
          }
        });
```

Since the `Observable.range` is just an example you can use `PublishSubject` as emitter of the single integer events.

**Comment by codebymikey** [comment]
Hi, thanks again for yet another relatively quick response! 

I wasn't suggesting that you pull RxJava in as a dependency, I was suggesting a generic [`debounce`](http://stackoverflow.com/questions/4742210/implementing-debounce-in-java) implementation which ensures that the `setText()` method is only called once every `n` seconds to avoid blocking the main thread.

The general idea being that it adds the text into the buffer but only calls `setText()` after a given interval has passed. As I don't think it's efficient to try calling `setText()` 1000 times in a second if you don't have to.

The block of RxJava code I included wasn't a proposed solution (which is probably where the Rx confusion came from) and was just some arbitrary code I was messing around with which lead me to open the original issue. I could've chosen not to log it, or have used a concat like you suggested, but I just didn't like the idea that a Timber logging Interface would be limited by the number of calls potentially made to it in a second.

I also don't see any major disadvantages to opening access to the `ConsoleBuffer` class/object as it allows developers to extend the class to modify it's behaviour to suit their needs without being blocked by new releases (kind of like the situation we're currently in). Do you have any reasons for not wanting to implement it which I may be missing?

**Comment by codebymikey** [comment]
Pseudo code for my proposed solution is to use an [`Handler`](https://developer.android.com/reference/android/os/Handler.html) in the `ConsoleBuffer` class making calls to `postDelayed` and `removeCallbacks`. kind of like:

```
Runnable runnable = new Runnable() {
    public void run() {
      textView.setText(buffer);
      waiting = false;
   }
};
Handler handler; // initialized somewhere else
long interval = 1000; // one second.
long lastCall = -1; 
boolean waiting = false;
boolean debounce = false;

 ConsoleBuffer printTo(TextView textView) {
      if(debounce) {
         // delay the logging
         if(lastCall != -1 && ((System.currentMillisecond() - lastCall) <= interval)) {
             if(!waiting) {
                  // run this after n seconds.
                  waiting = true;
                  handler.postDelayed(runnable, interval);
              }
         } else {
             // cancel existing setText runnables if any.
             handler.removeCallbacks(runnable);
             textView.setText(buffer);
         }
         lastCall = System.currentMillisecond();
      } else {
         // Use the default synchronized solution
         synchronized(lock) {
             textView.setText(buffer);
        }
      }

    return this;
  }
```

The `debounce` specific fields can be lazy-loaded only when `setDebounce` is called. This way any extra memory consumption is non-existent.

edit: Better yet, you can do away with the `Handler` and just do the `postDelayed` and `removeCallbacks` callbacks directly off the TextView object.

**Comment by jraska** [comment]
Two notes here: 
1.  Reason why I am not keen to add this to library is that intent of Console is only create static bridge for setText method to provide convenient way for client to log some debug messages. Solution you are proposing brings some additional complexity to maintain, which can make sense, but
2. What we want here is not `debounce` behaviour, it is `buffer` behaviour, because when there will be lot of log messages now for long time, nothing will be printed at all, because each print will be just postponed.

Feel free to submit PR for the buffer behaviour, because I don't have time now to look into it. Please remember we need test for it as well.

**Comment by codebymikey** [comment]
I got my terminologies mixed up. I meant `buffer` not `debounce`.

I'll look into it implementing a `buffer` behaviour and submit it as a PR.
